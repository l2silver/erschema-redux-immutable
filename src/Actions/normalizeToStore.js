// @flow
import normalize from 'erschema'
import {batchActions} from 'redux-batched-actions'
import {get} from 'lodash';
import {entityActions, relationshipActions} from '../actions-handlers'


import type {$schemaWithRelationshipsArray, $entitySchemaWithRelationshipsArray} from 'erschema/types'


export default function normalizeToStore(entity: Object, name: string, schema: $schemaWithRelationshipsArray, firstSchema?: $entitySchemaWithRelationshipsArray, options?: Object = {}){
  const {entities: allEntities, relationships: allRelationships} = normalize(entity, name, schema, firstSchema)
  const entitiesForStore = Object.keys(allEntities).reduce((finalResult, entityName)=>{
    const entities = allEntities[entityName]
    const entitiesForIndex = Object.keys(entities).reduce((finalResult, entityId)=>{
      finalResult.push(entities[entityId])
      return finalResult
    }, [])
    finalResult.push(entityActions.index(entityName, entitiesForIndex))
    return finalResult
  }, [])
  const relationshipsForStore = Object.keys(allRelationships).reduce((allRelationshipActions, entityName)=>{
    const relationships = allRelationships[entityName]
    const relationshipsForIndex = Object.keys(relationships).reduce((finalResult, relationshipName)=>{
      const relationship = relationships[relationshipName]
      const idValuePairs = Object.keys(relationship).reduce((finalResult, entityId)=>{
        finalResult.push({id: entityId, value: relationship[entityId]})
        return finalResult
      }, [])
      
      const relationshipActionName = get(options, `relationships.${entityName === 'pages' ? idValuePairs[0].id : entityName}.${relationshipName}.concat`) ? 'concatRelationship' : 'index'
      finalResult.push(relationshipActions[relationshipActionName](entityName, {name: relationshipName, idValuePairs}))
      return finalResult
    }, allRelationshipActions)
    return allRelationshipActions
  }, [])
  return batchActions([...entitiesForStore, ...relationshipsForStore])
}

export function indexNormalizeToStore(entities: Object[], name: string, schema: $schemaWithRelationshipsArray, firstSchema?: $entitySchemaWithRelationshipsArray, options?: {}){
  return batchActions(
    entities.map((entity)=>{
      return normalizeToStore(entity, name, schema, firstSchema)
    })
  )
}