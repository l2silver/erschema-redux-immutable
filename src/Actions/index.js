// @flow
import {createAction} from 'redux-actions'
import {retypeAction} from 'redux-retype-actions'
import resourceActions, {generateActionName as actionifyName} from 'resource-action-types'
import {entityActions, relationshipActions} from '../actions-handlers'
import normalizeToStore, {indexNormalizeToStore} from './normalizeToStore'
import type {$schema} from 'erschema/types'

type $$id = string | number

const getRelatedName = (name)=>`GET_RELATED_${actionifyName(name)}`;
const concatRelatedName = (name)=>`CONCAT_RELATED_${actionifyName(name)}`;

export default class Actions {
  entities: {
    create: (ent: Object) => any,
    update: (ent: Object) => any,
    remove: (id: $$id) => any,
    get: (ent: Object) => any,
    index: (ents: Object[]) => any,
    getRelated: (id: $$id, relationshipName: string, ents: Object[]) => any,
    getAdditionalEntityProperties: (id: $$id, entityName: string, entity: Object) => any,
    concatRelated: (id: $$id, relationshipName: string, ents: Object[]) => any,
  };
  relationships: Object;
  name: string;
  constructor (schema: $schema, name: string) {
    this.name = name
    this.entities = {
      create: (entity) => entityActions.create(name, entity),
      update: (entity) => entityActions.update(name, entity),
      remove: (id) => entityActions.remove(name, id),
      get: (entity) => retypeAction(resourceActions.get(name), normalizeToStore(entity, name, schema)),
      index: (entities) => retypeAction(resourceActions.index(name), indexNormalizeToStore(entities, name, schema)),
      getRelated: (id, relationshipName, entities) => {
        if (entities instanceof Error) {
          return retypeAction(getRelatedName(name), entityActions.get(name, entities))
        }
        return retypeAction(getRelatedName(name), normalizeToStore({id, [relationshipName]: entities}, name, schema))
      },
      getAdditionalEntityProperties: (id, entityName, entity) => {
        if (entity instanceof Error) {
          return entityActions.get(entityName, entity)
        }
        return retypeAction(`GET_ADDITIONAL_ENTITY_PROPERTIES_${actionifyName(name)}`, normalizeToStore({id, ...entity}, name, schema))
      },
      concatRelated: (id, relationshipName, entities) => {
        if (entities instanceof Error) {
          return entityActions.get(name, entities)
        }
        return retypeAction(
          `${concatRelatedName(name)}_${actionifyName(name)}`,
          normalizeToStore({id, [relationshipName]: entities}, name, schema, undefined,
          {
            relationships: {
              [name]: {
                [relationshipName]: {
                  concat: true
                }
              }
            }
          }))
      }
    }
  }
  createAction (name: string, payload: Object) {
    return createAction(name)(payload)
  }
}

export class PageActions {
  entities: {
    create: (ent: Object) => any,
    update: (ent: Object) => any,
    remove: () => any,
    get: (ent: Object) => any,
    getRelated: (relationshipName: string, ents: Object[]) => any,
    concatRelated: (relationshipName: string, ents: Object[]) => any,
  };
  constructor (schema: $schema, firstSchema: $schema, name: string) {
    this.entities = {
      create: (entity) => entityActions.create('pages', {...entity, id: name}),
      update: (entity) => entityActions.update('pages', {...entity, id: name}),
      remove: () => entityActions.remove('pages', name),
      get: (entity) => retypeAction(
        `${resourceActions.get('pages')}_${actionifyName(name)}`,
        normalizeToStore(entity, 'pages', schema, firstSchema[name])
      ),
      getRelated: (relationshipName, entities) => {
        if (entities instanceof Error) {
          return entityActions.get(name, entities)
        }
        return retypeAction(
          `${getRelatedName(name)}_${actionifyName(relationshipName)}`,
          normalizeToStore({[relationshipName]: entities}, 'pages', schema, firstSchema[name]))
      },
      concatRelated: (relationshipName, entities) => {
        if (entities instanceof Error) {
          return entityActions.get(name, entities)
        }
        return retypeAction(
          `${concatRelatedName(name)}_${actionifyName(name)}`,
          normalizeToStore({[relationshipName]: entities}, 'pages', schema, firstSchema[name],
          {
            relationships: {
              [name]: {
                [relationshipName]: {
                  concat: true
                }
              }
            }
          }))
      }
    }
  }
  createAction (name: string, payload: Object) {
    return createAction(name)(payload)
  }
}
