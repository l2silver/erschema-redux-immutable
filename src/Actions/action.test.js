// @flow
jest.mock('redux-batched-actions', ()=>({batchActions: (action)=>action}))
jest.mock('redux-retype-actions', ()=>({retypeAction: (name, action)=>action}))
jest.mock('erschema-action-handlers', ()=>({
  entityActions: {
    index: (entityName, entities)=>({entityName, entities})
  },
  relationshipActions: {
    index: (entityName, relationships)=>({entityName, relationships})
  }
}))
import Actions from './'
import {standardizeEntity} from 'erschema'

const getId = ()=>Math.floor((Math.random()*100000))
describe('Action', function() {
  const schema = {
    users: standardizeEntity({
      Model: class {},
      properties: ['name', 'id'],
      relationships: {
        manyRelationships: {
          friends: [{}]
        },
        monoRelationships: {
        },
      },
    }),
    friends: standardizeEntity({
      Model: class {},
      properties: ['name', 'id']
    })
  };
  class ActionWithSchema extends Actions {
    constructor(name){
      super(schema, name)
    }
  }
  const action = new ActionWithSchema('users')
  describe('creates an action instance with access to', function(){
    it('get', function() {
      const entity = {id: 1, name: 'John'}
      const results = action.actions.get({...entity, friends: []})
      expect(results).toMatchSnapshot()
    })
  })
})