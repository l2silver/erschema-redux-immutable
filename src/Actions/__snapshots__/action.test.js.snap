// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action creates an action instance with access to get 1`] = `
Array [
  Object {
    "entities": Array [
      Object {
        "id": 1,
        "name": "John",
      },
    ],
    "entityName": "users",
  },
  Object {
    "entityName": "users",
    "relationships": Object {
      "idValuePairs": Array [
        Object {
          "id": "1",
          "value": Array [],
        },
      ],
      "name": "friends",
    },
  },
]
`;
